proposals/lip-0057.md

```diff
@@ -727,12 +727,13 @@ stakeTransactionParams = {
 
 Let `trsParams = decode(stakeTransactionParams, trs.params)` for the given transaction `trs`. Then the following verification steps are performed:
 
-* The function `validateObjectSchema(stakeTransactionParams, trsParams)` muss pass without exception.
+* The function `validateObjectSchema(stakeTransactionParams, trsParams)` must pass without exception.
 * `trsParams.stakes` has at most `2 * MAX_NUMBER_STAKING_SLOTS` elements. The reason of choosing this bound on the size is to allow a staker to unstake all current validators (which are at most `MAX_NUMBER_STAKING_SLOTS`) and stake with `MAX_NUMBER_STAKING_SLOTS` new ones in the same transaction.
 * A given `validatorAddress` is included in at most one stake from the list of stakes (regardless of the associated amounts).
 * For all stakes included in `trsParams.stakes`, we have:
   * `amount` value is a multiple of `BASE_STAKING_AMOUNT`, i.e., `amount % BASE_STAKING_AMOUNT == 0`. For the Lisk mainchain, where `BASE_STAKING_AMOUNT = 10^9` and `TOKEN_ID_POS` is the token ID of the LSK token, this corresponds to multiples of 10 LSK.
   * `amount != 0`.
+* Let `senderAddress` be the address of the user sending the transaction (i.e., the staker). Let `totalUpstake` be the sum of all positive amounts of the `trsParams.stakes` array, i.e., the total amount that the sender is staking in this transaction. Then the following should hold: `Token.getAvailableBalance(senderAddress, TOKEN_ID_POS) >= totalUpstake`.    
```

Comment:
I think you should still subtract the negative balance, as those are then again available to stake. In that case tho you need extra check to be sure that the unstake amount is correct. Also, why don't we use pseudocode here?

Reply:
I agree with @ricott1 here.

@gkoumout what you wrote in our discussion 4 days ago seems to be more correct:

```typescript
availableBalance + totalUnstakedAmount >= totalStakedAmount
```

Or, using the semantics from this document:

```typescript
Token.getAvailableBalance(senderAddress, TOKEN_ID_POS) >= totalUpstake - totalDownstake
```

The system should allow a staker to move his complete stake to another validator in 1 transaction, even when staker's balance is lower than the amount already staked, right?

Reply:
I also thought like that in the beginning, but this is actually not true for the current protocol: Once unstake, the amount remains locked for the locking period and will be unlocked later using the unlock command. Therefore, just after un-staking, the amount is not available; so I guess the check is correct.

Reply:
@ricott1 I did not write pseudocode to have it faster, I can refactor this later.